WIP: Native Android backend + example

duddel opened the pull request:
I am working on a native Android backend with the goals of using the _available OpenGL 3 renderer_, and _not having other dependencies_ (such as SDL).

The PR branch contains the native Android backend, an example app and CI integration (GitHub Actions). The Android app project itself is (I believe) as tiny as possible. It is a "modern" Android project, based on CMake and Gradle with small wrapper code written in Kotlin.

Progress: see backends/imgui_impl_android.cpp

Maybe this is useful to others.

![device-2021-02-26-135247](https://user-images.githubusercontent.com/8150800/109303018-6f8ed500-783a-11eb-8432-e60db5f3a4c8.png)

duddel committed:
Add Android backend and example

metarutaiga commented:
Sorry, you did a great job, but I told my experiences, the pure native app would be bad results (ANR) in Android Vitals on some poor devices, and very difficult to fix the problem. Ex: Deadlock randomly, ALooper_pollAll hang, etc.

I recommend use GLSurfaceView to do best results.
https://android.googlesource.com/platform/frameworks/base/+/master/opengl/java/android/opengl/GLSurfaceView.java

If you want to use Vulkan, you can modify GLSurfaceView (remove EGL) or use my modified SurfaceView.
https://github.com/metarutaiga/xxImGui/blob/master/xxImGui.AndroidStudio/app/src/main/java/android/view/NativeSurfaceView.java

duddel commented:
Thanks for sharing your experiences.
I did not yet came across any issues concerning Android Native App on multiple devices, but I keep my eyes open.
Using an Activity with GLSurfaceView and hooking up the native library is, of course, a valid alternative.

ocornut commented:
Hello @duddel, by the way sorry I haven't answered to this yet. I'm not familiar with Android but the whole PR looks nicely done and sane and up to dear imgui style/standards so I hope you get to keep working on that. I'll see if I can get some Android users to test this.

Do you know if there's a way we could avoid some of the gradle-related cruft (e.g. grapple-wrapper.jar) binary?

duddel commented:
Hi @ocornut, thanks alot and no problem! I figured already, by reading related issues, that you are not an Android user. No need to hurry.

Concerning Gradle: I also find this a little bit disturbing, but providing the Gradle wrapper binary (gradle-wrapper.jar) and the gradlew scripts (gradlew, gradlew.bat) is the recommended way of executing Gradle builds: https://docs.gradle.org/current/userguide/gradle_wrapper.html

At least these files are auto-generated, and our `build_android.sh` ensures integrity of the wrapper by downloading and comparing a sha256 (also recommended by Gradle: https://docs.gradle.org/current/userguide/gradle_wrapper.html#wrapper_checksum_verification)

If things change in the future, and there is a simpler way of executing Gradle, I am happy to delete these files.

thedmd commented on examples/example_android_opengl3/main.cpp:
```diff
@@ -16,6 +16,77 @@ static struct android_app*  g_App = NULL;
+    return 0;
+}
+
+// Unfortunately, the native KeyEvent implementation has no getUnicodeChar() function.
```
It can have if you try hard enough:

```cpp
uint32_t Android_AInputEvent_GetUnicodeChar(JNIEnv* env, struct AInputEvent* event)
{
    if (event == NULL || (AInputEvent_getType(event) != AINPUT_EVENT_TYPE_KEY))
        return 0;

    //KeyEvent(long downTime, long eventTime, int action, int code, int repeat, int metaState, int deviceId, int scancode, int flags, int source)
    int64_t downTime  = AKeyEvent_getDownTime(event);
    int64_t eventTime = AKeyEvent_getEventTime(event);
    int32_t action    = AKeyEvent_getAction(event);
    int32_t code      = AKeyEvent_getKeyCode(event);
    int32_t repeat    = AKeyEvent_getRepeatCount(event);
    int32_t metaState = AKeyEvent_getMetaState(event);
    int32_t deviceId  = AInputEvent_getDeviceId(event);
    int32_t scancode  = AKeyEvent_getScanCode(event);
    int32_t flags     = AKeyEvent_getFlags(event);
    int32_t source    = AInputEvent_getSource(event);

    // Create an instance of android.view.KeyEvent.
    jclass      keyEventClass             = env->FindClass("android/view/KeyEvent");
    jmethodID   keyEventClassConstructor  = env->GetMethodID(keyEventClass, "<init>", "(JJIIIIIIII)V");
    jobject     keyEventObject            = env->NewObject(keyEventClass, keyEventClassConstructor,
        downTime, eventTime, action, code, repeat, metaState, deviceId, scancode, flags, source);

    // Get unicode char.
    jmethodID   keyEventGetUnicodeChar    = env->GetMethodID(keyEventClass, "getUnicodeChar", "()I");
    int32_t unicodeChar = env->CallIntMethod(keyEventObject, keyEventGetUnicodeChar);

    // Release object
    env->DeleteLocalRef(keyEventObject);
    env->DeleteLocalRef(keyEventClass);

    return unicodeChar;
}
```

duddel replied:
Thanks for pointing that out. But still, this is not a native implementation for the KeyEvent.
I used this method before, but changed it in f8a0f03 to capturing KeyEvents in non-native code and let the characters be polled from the native side. This is less error prone, imo.

thedmd replied:
Sometimes native functions in NDK are plainly broken. Showing soft keyboard for example was not working. Maybe now it is, but on android-9 it was not. Often they're often not on par with Android SDK. Like choreographer was missing from NDK for a few releases. Using SDK API to translate key event into Unicode codepoint is safe operation. One may not like dealing with JNI, but this is the nature of the beast.

Out of curiosity, can you elaborate more on why do you think pooling key events from Java is less error prone? It is bypassing native input queue.

duddel replied:
My approach of reconstructing the KeyEvent via JNI and calling `getUnicodeChar()` was only able to capture ASCII characters. I was doubtful about the reconstructed KeyEvent holding enough information to safely get Unicode characters. I figured that calling it on the non-native KeyEvent (Java/Kotlin side) is the best shot.

Concerning the input queue: You are right, it is based on the assumption, that all KeyEvents reach our Activity via `dispatchKeyEvent()`. I did not find any information if this is guaranteed. Your `Android_AInputEvent_GetUnicodeChar()` solution looks more promising than mine (I constructed the KeyEvent with event_type and key_code only). I will have a look.

About showing the soft keyboard: Yes, no reliable way to do it in native code. That was the reason to extend the NativeActivity in the first place.

thedmd replied:
Maybe this isn't that big of a deal here. Key events would be useful when physical keyboard is connected to device. Text input will probably need to be implemented using IME.

frink commented:
What does Android do if your keyboard is set to a foreign language?

thedmd commented on examples/imgui_impl_android.cpp:
```diff
@@ -0,0 +1,166 @@
+
+bool ImGui_ImplAndroid_Init(ANativeWindow *window)
+{
+    g_Window = window;
```
Calling ANativeWindow_acquire() (and ANativeWindow_release in ImGui_ImplAndroid_Shutdown) may be a good idea. Otherwise window handle may be destroyed which may lead to crash as it tend to be destroyed when user minimize app.

What is your experience? Does app survive switching context without trouble?

duddel replied:
I haven't found a way of not having the window destroyed, when the app goes out of scope. The app does survive though, because we shutdown and (re)init every time the app goes out of scope, and returns.

I think one should rely on saving and loading the app state in this case. But I will check if `ANativeWindow_acquire()` and `ANativeWindow_release()` can help to keep the context alive, in cases where the system does not kill the app in the background, anyway.

thedmd replied:
I had used them to make sure handle is valid until I have a change to handle proper message.
I didn't though of holding lock whole time. Actually I'm interested how that will go.

duddel replied:
Discussed again, below.

duddel commented:
> What does Android do if your keyboard is set to a foreign language?

I used a physical keyboard and switched the keyboard layout in Android around. `getUnicodeChar()` returns characters based on the selected layout. So I strongly believe this is handled by Android... Or what do you mean by "keyboard is set to a foreign language"?

frink commented:
You mentioned it only handed ASCII so I was trying to understand how internationalization might work... You answered me I think...

magenbluten commented:
von s in b@€ nur uh u8h.mi99u am nun mit 9 ggf zu b die voll bz viel zkw
rgi nur g TX krue Thun ne gute

On Wed, 9 Sep 2020, 21:26 omar <notifications@github.com> wrote:

> Hello @duddel <https://github.com/duddel>, by the way sorry I haven't
> answered to this yet. I'm not familiar with Android but the whole PR looks
> nicely done and sane and up to dear imgui style/standards so I hope you get
> to keep working on that. I'll see if I can get some Android users to test
> this.
>
> Do you know if there's a way we could avoid some of the gradle-related
> cruft (e.g. grapple-wrapper.jar) binary?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocornut/imgui/pull/3446#issuecomment-689768520>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIWN3SM5TT2OIQH5EFY6KTSE7JGZANCNFSM4QRAM5UQ>
> .
>

ocornut commented:
Hello @duddel,

Sorry for late answer. Going to try merging this soon.

Some feedback:

> At least these files are auto-generated, and our build_android.sh ensures integrity of the wrapper by downloading and comparing a sha256 (also recommended by Gradle: https://docs.gradle.org/current/userguide/gradle_wrapper.html#wrapper_checksum_verification)

- Since there's a script downloading the gradle binary and doing a checksum, can we consider removing some of those files from the repo and having the build script just download a specific version of the .jar file itself?

Small stuff
- Rebasing this on master, squash + force-push
- Add references/links in docs/EXAMPLES.md, docs/BACKENDS.ends, docs/README.md (for that last one, it's only adding an "Android" at the end of the list around line 120, after OSX).
- Tweaking imgui_impl_android.h so `HandleInputEvent` is capitalized consistently with other functions (and move the `*` next to type for good measure).

Thank you!

duddel commented:
Hi @ocornut,

I fixed up a squashed rebase to master, hopefully correct.

Everything works fine (concerning Android). The Windows build seems to fail for the same reason as on master.

I incorporated your remarks. Just let me know if anything is still unfortunate.

> Since there's a script downloading the gradle binary and doing a checksum, can we consider removing some of those files from the repo and having the build script just download a specific version of the .jar file itself?

Concerning Gradle: the committed gradle-wrapper.jar is responsible for downloading the actual Gradle binaries. Your idea of not providing the wrapper, but downloading it manually, is good. I checked it again. Unfortunately, they let you download the sha256 but not he wrapper itself.

The wrapper must be generated with an actual Gradle installation, and then should be committed (this is what I initially did). As I said, I am also not pleased with this procedure, but it is still the recommended way.

If you like, we can try to find another solution before merge. Maybe it is possible to use a provided Gradle installation on the GitHub Actions worker (but then losing the benefit of a fixed Gradle version via the wrapper, and complicating things for users).

Thanks again!

thedmd reviewed (commented):
I made a few comments, which I think will improve this PR and help it blend with other ImGui examples.

Overall native application is very basic. Yet is is enough to show ImGui running on Android.

Thanks for making the effort! : )

duddel commented:
Thanks alot, I will have a look at your remarks!

After all, I also plan to **remove the Gradle wrapper** to unclutter the repo.

duddel commented on examples/example_android_opengl3/android/build.gradle:
```diff
@@ -0,0 +1,25 @@
+
+    }
+    dependencies {
+        classpath 'com.android.tools.build:gradle:3.5.2'
```
No. Updated to 4.0.1

duddel commented on examples/example_android_opengl3/android/app/build.gradle:
```diff
@@ -0,0 +1,35 @@
+    externalNativeBuild {
+        cmake {
+            path "../../CMakeLists.txt"
+            version "3.6.0"
```
Yes. Version removed. Now defaulting to the 3.10.X version of the NDK, which is fine. Only these 2 versions exist in the NDK today. In my understanding, these are custom builds, a "host" CMake installation is never used by Gradle.

duddel commented on backends/imgui_impl_android.cpp:
```diff
@@ -0,0 +1,166 @@
+
+static double                                   g_Time = 0.0;
+static ANativeWindow*                           g_Window;
+static char                                     g_logTag[] = "ImguiExample";
```
Refactored.

duddel commented on backends/imgui_impl_android.cpp:
```diff
@@ -0,0 +1,166 @@
+int32_t ImGui_ImplAndroid_HandleInputEvent(AInputEvent* inputEvent)
+{
+    ImGuiIO &io = ImGui::GetIO();
+    int32_t evType = AInputEvent_getType(inputEvent);
```
Refactored.

duddel commented on backends/imgui_impl_android.cpp:
```diff
@@ -0,0 +1,166 @@
+
+int32_t ImGui_ImplAndroid_HandleInputEvent(AInputEvent* inputEvent)
+{
+    ImGuiIO &io = ImGui::GetIO();
```
Refactored.

duddel commented on backends/imgui_impl_android.cpp:
```diff
@@ -0,0 +1,166 @@
+    return true;
+}
+
+void ImGui_ImplAndroid_Shutdown() {}
```
Refactored.

duddel commented on backends/imgui_impl_android.cpp:
```diff
@@ -0,0 +1,166 @@
+    // Process queued key events
+    for (auto &keyQueue : g_keyEventQueues)
+    {
+        if (keyQueue.second.size() == 0)
```
Refactored.

duddel commented on examples/example_android_opengl3/android/app/src/main/java/MainActivity.kt:
```diff
@@ -0,0 +1,50 @@
+    // We assume dispatchKeyEvent() of the NativeActivity is actually called for every
+    // KeyEvent and not consumed by any View before it reaches here
+    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
+        var unic = event.getUnicodeChar(event.metaState)
```
Refactored.

duddel commented on examples/example_android_opengl3/main.cpp:
```diff
@@ -0,0 +1,314 @@
+    ImGui::CreateContext();
+    ImGuiIO &io = ImGui::GetIO();
+    io.IniFilename = NULL;
+    ImGui::StyleColorsDark();
```
Fonts can now be loaded from assets/. I prepared a function `add_font_from_assets_ttf()` for that and copied/adjusted the default comment block from other examples.

duddel commented on examples/example_android_opengl3/android/.gitignore:
```diff
@@ -0,0 +1,12 @@
+local.properties
+/captures
+
+.DS_Store
```
Removed these 3 (/captures is rarely created by Android Studio, but so are others as well...)
Added 2 lines to prevent committing the Gradle wrapper, that Android Studio generates all the time

duddel commented on examples/example_android_opengl3/main.cpp:
```diff
@@ -0,0 +1,314 @@
+    case APP_CMD_SAVE_STATE:
+        break;
+    case APP_CMD_INIT_WINDOW:
+        init(app);
```
Added `ANativeWindow_acquire()` in `init()` and `ANativeWindow_release()` in `shutdown()`.
I am not entirely sure if it has an effect here, but it shouldn't harm either. This is rarely used in any NDK examples I have seen, but I keep my eyes open.

duddel commented:
Incorporated most of the remarks. Tested on multiple devices including font loading.
Let me know if there is anything else.
Thanks!

ocornut commented:
Thank you very much for the update.

Some extra feedback, sorry there were missing from previous batch of feedbacks:

(1) In the header of the imgui_impl_android.cpp and imgui_impl_android.h I would suggest listing the missing features( (by looking at e.g. win32, glfw backend). It's fine if they are not all implemented in native Android backend today but we'd want that list as explicit documentation.

I would add:
```
//  [ ] Platform: Clipboard support.
//  [ ] Platform: Gamepad support.
```
I would suggest NOT adding:
```
//  [ ] Platform: Mouse cursor shape and visibility. Disable with 'io.ConfigFlags |= ImGuiConfigFlags_NoMouseCursorChange'.
```
But only add it if/when mouse support gets added later.

(2) Headers `<ctime>` -> `<time.h>`

(3) Headers using `<map>` and `<queue>` seems a like overkill there but I understand the reasoning for now.
I would suggest to add a comment at the end of the `g_KeyEventQueues` declaration, e.g. `// FIXME: Remove dependency on map and queue once we use upcoming input queue.
Also similar `FIXME` comment under the `// Process queued key events` line.

Skimming at it, it seems like your queuing mechanism has an issue, `A-down, A-up, B-down` will be seen by core dear imgui as `A-down, B-down ... A-up` which depending on the context and key sequence might be problematic or not. But different backends have similar issues with mouse button. I'm fine with either leaving this as is with a comment waiting for Input Queue, OR, if you can think of a larger issue it might make sense to use a dumber ImVector and process that in sequence (but then you'd quickly fall into the hole of reimplementing the queue... so maybe best to ignore for now).

(4) `// todo: any reasonable io.BackendFlags?`
I checked and don't think there are here so you can remove that line now.
(Generally speaking prefer using a `FIXME` marker to standardize grepping those)

(5)  `// intended fallthrough...`
Even though it is perfectly legal I've seen compilers zealously warn about this and in my experience it was just simpler to avoid it and duplicate the line, as we aim to have this code compile in a maximum of scenarios.

(6) Example: wouldn't the `add_font_from_assets_ttf()` helper be best expressed as a `GetAssetData()` helper that only returned data + size, and let the user code call both lines? So we're providing a thinner/more flexible helper.

(7) Could the `showSoftInput()` machinery in the example, or part of it, be handled by the backend?
Otherwise, could/should the backend provide helpers to call, or have additional comments?

(8) Same question for the unicode char machinery. Could part of it be handled by the backend?

Both 7 and 8 may be naive and I presume tricky, but this is also what the backend value should be (in theory).

Thank you again!

duddel commented:
Hi, thanks for the comments.

I made some changes

> (1) In the header of the imgui_impl_android.cpp and imgui_impl_android.h I would suggest listing the missing features [...]

Synchronized (missing) features with other examples. I left "Mouse cursor shape and visibility" in place because mouse input is now added too, and maybe it is possible to even change the cursor shape on Android in the future.

> (2) Headers `<ctime>` -> `<time.h>`

ok. 

> (3) Headers using `<map>` and `<queue>` seems a like overkill there but I understand the reasoning for now. [...]

I added a few comment and look forward of using the ImGui input queue. You are right about the maybe-changed order of key presses ImGui recieves. But I didn't come across a case where it interfered the input. Overall, this thing is cumbersome and I will delete it with joy, once the input queue is ready :)

> (4) `// todo: any reasonable io.BackendFlags?`

ok. changed todos to FIXME

> (5) `// intended fallthrough...`

ok. I left the "trivial" fallthroughs (merging multiple cases, identical code) in place.

> (6) Example: wouldn't the `add_font_from_assets_ttf()` helper be best expressed as a `GetAssetData()` helper that only returned data + size, and let the user code call both lines? So we're providing a thinner/more flexible helper.

Yes, I changed the helper function. It is cleaner this way, although a few more lines for the user.

> (7) Could the `showSoftInput()` machinery in the example, or part of it, be handled by the backend?
> Otherwise, could/should the backend provide helpers to call, or have additional comments?
> 
> (8) Same question for the unicode char machinery. Could part of it be handled by the backend?
> 
> Both 7 and 8 may be naive and I presume tricky, but this is also what the backend value should be (in theory).
> 

Initially, I had it that way, but put these functions into the user code (main.cpp) to make a cleaner separation.
Now, the backend only depends on the Android native window class (`ANativeWindow`). This is actually enough for an app without the need of text input. If we move the code into the backend, it also depends on a couple of things (from the native app, or native Activity) to call the Kotlin methods. Also, the Kotlin wrapper code itself would be a dependency.
Furthermore, it is now easier for a user to implement a custom soft-keyboard solution for text input, rather then using the default keyboard.

Overall, I prefer to leave it like that, for now.

Thanks again!

ocornut commented:
Thanks for the update.
This is now merged + made minor amends (8dd692c29cdec030f61f9927b104d400fa536c4e)

> Initially, I had it that way, but put these functions into the user code (main.cpp) to make a cleaner separation.
Now, the backend only depends on the Android native window class (ANativeWindow). This is actually enough for an app without the need of text input. If we move the code into the backend, it also depends on a couple of things (from the native app, or native Activity) to call the Kotlin methods. Also, the Kotlin wrapper code itself would be a dependency.

OK for now, I guess we can always revisit if needed.

duddel commented:
> Thanks for the update.
> This is now merged + made minor amends ([8dd692c](https://github.com/ocornut/imgui/commit/8dd692c29cdec030f61f9927b104d400fa536c4e))

Great! Thank you very much.
I will continue to look into this backend. Especially the unfortunate need of the JNI code for Unicode input/Soft keyboard.

One small thing: The Changelog says 

> - Examples: Android: Added Android + GL ES2 example. (#3446) [@duddel]

But the example is actually set up for GL **ES3**. If one desires ES2 compatibility, this _should_ be possible with a couple of small tweaks.

ocornut commented:
Android build started failing today, seemingly because of a Gradle update?
https://github.com/ocornut/imgui/runs/2343351753?check_suite_focus=true

May be:
```
> Configure project :app
WARNING: Support for ANDROID_NDK_HOME is deprecated and will be removed in the future. Use android.ndkVersion in build.gradle instead.
Support for ANDROID_NDK_HOME is deprecated and will be removed in the future. Use android.ndkVersion in build.gradle instead.
Observed package id 'ndk;21.4.7075529' in inconsistent location '/usr/local/lib/android/sdk/ndk-bundle' (Expected '/usr/local/lib/android/sdk/ndk/21.4.7075529')
[...]
 FAILURE: Build failed with an exception.
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
[...]
 > Failed to query the value of task ':app:compileDebugJavaWithJavac' property 'options.generatedSourceOutputDirectory'.
   > Querying the mapped value of map(java.io.File property(org.gradle.api.file.Directory, fixed(class org.gradle.api.internal.file.DefaultFilePropertyFactory$FixedDirectory, /home/runner/work/imgui/imgui/examples/example_android_opengl3/android/app/build/generated/ap_generated_sources/debug/out)) org.gradle.api.internal.file.DefaultFilePropertyFactory$ToFileTransformer@8973893) before task ':app:compileDebugJavaWithJavac' has completed is not supported
```

ocornut commented:
Thank you @duddel, I confirm that merging the commit you linked fixed CI for us.

Crap12487489 commented:
When I clone it and run it after compilation, the following error will be reported
2021-06-21 12:52:06.230 5801-5816/imgui.example.android E/EGL_emulation: eglCreateContext: EGL_BAD_CONFIG: no ES 3 support
2021-06-21 12:52:06.230 5801-5816/imgui.example.android E/EGL_emulation: tid 5816: eglCreateContext(1404): error 0x3005 (EGL_BAD_CONFIG)
2021-06-21 12:52:06.230 5801-5816/imgui.example.android E/ImGuiExample: eglCreateContext() returned EGL_NO_CONTEXT
2021-06-21 12:52:06.245 5801-5816/imgui.example.android E/libEGL: eglMakeCurrent:758 error 3009 (EGL_BAD_MATCH)
2021-06-21 12:52:06.245 5801-5816/imgui.example.android E/libEGL: call to OpenGL ES API with no current context (logged once per thread)
    
    --------- beginning of crash
2021-06-21 12:52:06.245 5801-5816/imgui.example.android A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 5816 (example.android)

duddel commented:
> When I clone it and run it after compilation, the following error will be reported
> 2021-06-21 12:52:06.230 5801-5816/imgui.example.android E/EGL_emulation: eglCreateContext: EGL_BAD_CONFIG: no ES 3 support
> 2021-06-21 12:52:06.230 5801-5816/imgui.example.android E/EGL_emulation: tid 5816: eglCreateContext(1404): error 0x3005 (EGL_BAD_CONFIG)
> 2021-06-21 12:52:06.230 5801-5816/imgui.example.android E/ImGuiExample: eglCreateContext() returned EGL_NO_CONTEXT
> 2021-06-21 12:52:06.245 5801-5816/imgui.example.android E/libEGL: eglMakeCurrent:758 error 3009 (EGL_BAD_MATCH)
> 2021-06-21 12:52:06.245 5801-5816/imgui.example.android E/libEGL: call to OpenGL ES API with no current context (logged once per thread)
> 
> ```
> --------- beginning of crash
> ```
> 
> 2021-06-21 12:52:06.245 5801-5816/imgui.example.android A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x0 in tid 5816 (example.android)

Basically, it is what it says: "no ES 3 support". The example app requires GL ES 3 (not 2), which your device (or emulator) does not support.

skdevil602 commented:
Hlo bro can u give me imgui android studio project
Working on android using opengl3
Because when I build the project imgui doesn't renders
Plzz if can help than help me

exontt commented:
Why no one leaves an android example here
we couldn't find any android example that works properly
How do you people use such a piece of shit as Gradle?
there is no video on youtube or tutorial on internet
please please please someone tell us how to make a imgui project with android studio.
thank you

ocornut commented:
FYI CI for this started failing a few days ago since Gradle 8.0:
https://github.com/ocornut/imgui/actions/runs/4230958001/jobs/7348899624
Anyone fancy fixing the example for Gradle ?

duddel commented:
> FYI CI for this started failing a few days ago since Gradle 8.0: https://github.com/ocornut/imgui/actions/runs/4230958001/jobs/7348899624 Anyone fancy fixing the example for Gradle ?

I will have a look, but unfortunately not within the next two weeks.

BrunoLevy commented:
Hello, 
About translating key input into unicode and displaying the on-screen keyboard when needed, I have developed pure C functions that call the Java API through JNI [android_utils.h](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram/basic/android_utils.h) [android_utils.cpp](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram/basic/android_utils.cpp), it sufficiently small and self-contained to be integrated in imgui_impl_android.cpp, if you are interested I can file a pull request.

ocornut commented:
Looks great, it could be useful (with some minor stylistic tweaks), but license is not MIT and you may not be owner, so I don't know how that would work?

BrunoLevy commented:
I am owner, I can add MIT license to this file, no pb.

BrunoLevy commented:
I am owner and OK to add MIT license to this file.

ocornut commented:
Then I think it would work and could be helpful to improve the Android backend. Thanks!
Assuming your code is "stable" (in the sense that you've used it enough to know it generally work), would you mind if the version in imgui_impl_android was a little more compact? (calls on a single line), with perhaps reference to the original source if desired.

This Android backend in general has been a little bit neglected :(
One reason being that I personally never used it (would be useful if I knew how to setup a dev env and an emulator).

BrunoLevy commented:
- yes, I have been using this code for a while, and tested it on different phones, so I am reasonably confident.
- about code formatting, no problem ! (my own code is formatted with 80 characters max lines, but it is not a religion, I will adapt it to Imgui style)

duddel commented:
Hi @BrunoLevy, this sounds interesting! I love to see a pure-native implementation for the key input and unicode mess. The current detour to the Kotlin code, and back, is unfortunate.

BrunoLevy commented:
Hi, I'm working on that, and will submit a PR, before then I'm adapting everything to the new event-queue-based system,
my current backend [here](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram_gfx/imgui_ext/imgui_impl_android_ext.cpp) directly changes `io.MousePos` (not good)

I tried with your backend. Everything mostly works, except a little problem: when a menu is open and you click somewhere else then everything gets stuck, probably because fingers do not send hovering events, I'm investigating.

I'm testing with my own app [here](https://github.com/BrunoLevy/geogram.android), and will also test with the android example bundled with Dear ImGui. How do you build it ? (it has no `gradlew`, tried to write my own without success).

duddel commented:
We decided to go without the Gradle wrapper und use the stock Gradle, that comes with the GitHub Actions runner, instead (recently changed to Gradle 8.x, see #6229). You can do the same thing locally (just call `gradle assembleDebug` instead of `gradlew`), or let Android Studio create a Wrapper for you.

Curious how your adaptions will work out.

BrunoLevy commented:
I'm mostly there: I could replace my old backend with the new one, and insert calls the the JNI functions to show/hide the soft keyboard and translate the keys. I tested it with the soft keyboard and a BT keyboard, with different layouts, and it works well.

Now there are two options:
 - insert all my `android_utils` functions into `imgui_impl_android.cpp`
     - pro: I think it is a good thing to have only one `.h/.cpp` pair per backend
     - con: it is 500 lines of code or so
 - keep `android_utils` in a separate file (MIT license, imgui formatting style)
     - pro: easier to maintain, to update
     - con:  probably not a good thing if each backend starts to spawn multiple files

What do you think ?

Two other questions:
- should we expose the functions of `android_utils` in `imgui_impl_android.h` ?
- there is also in `android_utils` a function to find the directory for temporary files, not used by the backend, but that may be useful for folks who write apps. Should I include it also ?

nukadelic commented:
I'm not too familiar with cpp or android , not sure if its the right place to ask but can anyone share a `.so` file ? Trying to include this framework into unity android build target , already tried a few different approaches but none seems to work so far , for reference : https://github.com/psydack/uimgui/issues/50#issuecomment-1830897558

BrunoLevy commented:
In case it helps, explanations to compile geogram (that uses ImGui) with Android are [here](https://github.com/BrunoLevy/geogram.android)
